### DISCUSSION TITLE
Uploading file with security dependency waits for full file to be loaded prior to authentication?

### DISCUSSION THREAD

**Discussion by piehld** [fastapi/fastapi#6406] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
# main.py (start up application with this)
from fastapi import Depends, FastAPI, HTTPException, status, File, UploadFile
from fastapi.security import HTTPBasic, HTTPBasicCredentials
import secrets

app = FastAPI()
security = HTTPBasic()

def get_current_username(credentials: HTTPBasicCredentials = Depends(security)):
    correct_username = secrets.compare_digest(credentials.username, "stanleyjobson")
    correct_password = secrets.compare_digest(credentials.password, "swordfish")
    if not (correct_username and correct_password):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect email or password",
            headers={"WWW-Authenticate": "Basic"},
        )
    return credentials.username

@app.post("/uploadfile/")
async def create_upload_file(file: UploadFile = File(...), username: str = Depends(get_current_username)):
    return {"filename": file.filename}


# uploadFile.py (to be put into separate file and run separately)
import requests
from requests.auth import HTTPBasicAuth
import time
import os

badAuth = HTTPBasicAuth(username="badstanley", password="badfish")
goodAuth = HTTPBasicAuth(username="stanleyjobson", password="swordfish")

smallFile = "smallFile.txt"
bigFile = "bigFile.txt"

if not os.path.exists(smallFile):
    with open(smallFile, "wb") as ofh:
        ofh.write(os.urandom(262144))  # generate 256 kb file

if not os.path.exists(bigFile):
    with open(bigFile, "wb") as ofh:
        ofh.write(os.urandom(268435456))  # generate 256 mb file

for file in [smallFile, bigFile]:
    print(file)
    for auth in [badAuth, goodAuth]:
        files = {"file": open(file, 'rb')}
        startTime = time.time()
        res = requests.post("http://0.0.0.0:8000/uploadfile", files=files, auth=auth)
        print(res.text, f"took {time.time() - startTime:.4f} seconds")
```

### Description
***Note:*** Basically the same as https://github.com/tiangolo/fastapi/issues/3165, but here I provide a full example and am curious if this is intended behavior or an opportunity for code improvement.

If I upload a file using an upload endpoint that depends on authentication, it seems that the full file is uploaded prior to checking whether the provided credentials are valid. E.g., if you start up the application below (`main.py`) and run the provided `uploadFile.py` script, you'll see that the time it takes to deny the large file upload request is significantly longer than the time to deny the small file upload request. To me, this indicates that the entire file is attempted to be read-in by the application prior to checking for authentication.

So I guess my question is three-fold:
1. Is this expected behavior?
2. Wouldn't it be more efficient to check authentication before spending the time reading in the entire upload file request
3. How would I implement a way to perform the upload in such a way? Using a middleware endpoint?

Thank you!

### Operating System

macOS

### FastAPI Version

0.78.0

### Python Version

4.9.7

**Comment by adriangb** [comment]
I wouldn't say this is expected or unexpected: it's an implementation detail.

I think not always but generally it would be best do do auth first. You can't change this behavior. If you are receiving a single file you could just stream the bytes instead of using an UploadFile.

**Comment by piehld** [comment]
Thanks, I ended up using a middleware instead. Appreciate the response and advice.

**Comment by tiangolo** [comment]
Thanks for the help @adriangb! 🙇 

And thanks for coming back to close the issue after finding a solution for your use case @piehld. 🍰 

> Sorry for the long delay! 🙈  I wanted to personally address each issue and they piled up through time, but now I'm checking each one in order.

**Comment by anubhav-sl** [comment]
> Thanks, I ended up using a middleware instead. Appreciate the response and advice.

What are you using ?
